source/common/config/grpc_subscription_impl.h

```diff
@@ -36,114 +32,28 @@ class GrpcSubscriptionImpl : public Config::Subscription<ResourceType>,
-    Protobuf::RepeatedPtrField<ProtobufTypes::String> resources_vector(resources.begin(),
-                                                                       resources.end());
-    request_.mutable_resource_names()->Swap(&resources_vector);
-    sendDiscoveryRequest();
-    stats_.update_attempt_.inc();
-  }
-
-  const std::string versionInfo() const override { return request_.version_info(); }
-
-  // Grpc::AsyncStreamCallbacks
-  void onCreateInitialMetadata(Http::HeaderMap& metadata) override {
-    UNREFERENCED_PARAMETER(metadata);
+    ads_subscription_.updateResources(resources);
   }
 
-  void onReceiveInitialMetadata(Http::HeaderMapPtr&& metadata) override {
-    UNREFERENCED_PARAMETER(metadata);
-  }
+  const std::string versionInfo() const override { return ads_subscription_.versionInfo(); }
 
-  void onReceiveMessage(std::unique_ptr<envoy::api::v2::DiscoveryResponse>&& message) override {
-    const auto typed_resources = Config::Utility::getTypedResources<ResourceType>(*message);
-    try {
-      callbacks_->onConfigUpdate(typed_resources);
-      request_.set_version_info(message->version_info());
-      stats_.update_success_.inc();
-      ENVOY_LOG(debug, "gRPC config update accepted: {}", message->DebugString());
-    } catch (const EnvoyException& e) {
-      ENVOY_LOG(warn, "gRPC config update rejected: {}", e.what());
-      stats_.update_rejected_.inc();
-      callbacks_->onConfigUpdateFailed(&e);
-    }
-    // This effectively ACK/NACKs the accepted configuration.
-    ENVOY_LOG(debug, "Sending version update: {}", message->version_info());
-    stats_.update_attempt_.inc();
-    request_.set_response_nonce(message->nonce());
-    sendDiscoveryRequest();
-  }
-
-  void onReceiveTrailingMetadata(Http::HeaderMapPtr&& metadata) override {
-    UNREFERENCED_PARAMETER(metadata);
-  }
-
-  void onRemoteClose(Grpc::Status::GrpcStatus status, const std::string& message) override {
-    ENVOY_LOG(warn, "gRPC config stream closed: {}, {}", status, message);
-    handleFailure();
-    stream_ = nullptr;
-  }
-
-  // TODO(htuch): Make this configurable or some static.
-  const uint32_t RETRY_DELAY_MS = 5000;
+  GrpcMuxImpl& grpcMux() { return grpc_mux_; }
 
 private:
-  void handleFailure() {
-    stats_.update_failure_.inc();
-    callbacks_->onConfigUpdateFailed(nullptr);
-    setRetryTimer();
-  }
-
-  std::unique_ptr<
-      Grpc::AsyncClient<envoy::api::v2::DiscoveryRequest, envoy::api::v2::DiscoveryResponse>>
-      async_client_;
-  const Protobuf::MethodDescriptor& service_method_;
-  Event::TimerPtr retry_timer_;
-  Protobuf::RepeatedPtrField<ProtobufTypes::String> resources_;
-  Config::SubscriptionCallbacks<ResourceType>* callbacks_{};
-  Grpc::AsyncStream<envoy::api::v2::DiscoveryRequest>* stream_{};
-  envoy::api::v2::DiscoveryRequest request_;
-  SubscriptionStats stats_;
+  GrpcMuxImpl grpc_mux_;
+  GrpcMuxSubscriptionImpl<ResourceType> ads_subscription_;
```

Comment:
I'm a little confused here. For a singleton gRPC subscription, how does this work? Does this somehow go to the null ADS?

Reply:
Or we are just using same code? If so rename grpc_subscription_?

Reply:
It uses its own instance of `GrpcMuxImpl`, with jut a single subscription on the channel. Will update variable name as suggested.